### ISSUE TITLE
http: Strategy for Authentication, Authorization, Audit

### ISSUE THREAD

**Issue by briantopping** [akka/akka#17667]
Akka HTTP is a "general toolkit for providing and consuming HTTP-based services". It's practically impossible to imagine a HTTP service that doesn't have minimal requirements for user authentication, which quickly leads to different users having different abilities within the system (administrative, paid, earned, etc). https://github.com/akka/akka/issues/16857 covers one method of authentication and the HTTP protocol specifies support for basic and digest authentication, but it's quickly becoming evident that passwords alone are not sufficient authenticators (most major consumer services offer two-factor using mobile phones nowadays).

So what is the best strategy for Akka HTTP? OAuth is an accepted feature request and there are several variants of it once one considers the different means of obtaining profile information. It covers authentication, but not authorization -- once my identity has been established, what can I do? Audit is the last leg of AAA, once I have permissions, how can the system uniformly record the use of those permissions? For instance, it would be desirable that HTTP routing could block unauthorized users, collect context parameters automatically when a user successfully enters a block, and flexibly log in either case.

With the maturity of OSS products like FreeIPA, it seems like the time has come to look beyond traditional in-app authentication and authorization, but smaller teams that can't afford that overhead also need to be able to "hack something out". The same is true for authorization.

I could keep yammering about this, but it's probably best to start with whether the team thinks this kind of discussion is in scope and go from there. Scope definition will help with whether this should be a built-in or a SPI, but I think if it's fuzzy for too long, apps will suffer when it is finally unified. It's also worth considering that teams that are using Akka have already signed up to "do it right", so there may be more interest in this than is initially considered.

**Comment by sirthias** [comment]
I think that the main feature that Akka HTTP has to offer is support for the most important authentication mechanisms that are typically employed with HTTP. Basic Auth and OAuth are probably the ones.
Digest Auth appears to have fallen out of favor, for good reasons.
Authentication is a more of a low-level feature, with established specifications, that makes sense for a library to provide in order to prevent users from having to rebuild the same logic every time.

Anything after that, i.e. authorization, audit and more, are less a concern of HTTP than of the application layer. Here the needs will differ significantly between applications.
As in spray Akka HTTP gives you simple support for authorization with the `authorize` directive, which, however, does very little that you cannot easily build yourself if you want.
My take is that with a flexible DSL like the routing DSL users have all the tools they need to easily define the authentication + audit + ... solution that is right for their use case.

What exactly would it be that you are looking for?

**Comment by briantopping** [comment]
One of my goals as an app developer are to not get locked in to an AA provider that has 80% of the solution or hasn't been keeping up with the latest "it" methods. It's a wide world out there, so making this stuff pluggable helps a lot. I think the toolkit should provide a solid SPI for authentication but not get locked in to API for specific providers. If you look at Spring Security, they've done a great job at providing a single API to dozens of different authentication schemes. SS is an add-in to the toolkit, not a part of it, so the add-in can be reved without waiting for the toolkit.

From a marketing perspective, having everything external means Akka is far less likely to be listed [here](https://www.us-cert.gov/ncas/alerts).

As a developer, my primary concerns are that I can test _for the current user_ having a Permission with a very lightweight call. As an architect, I don't want developers to be able to test on a Role, because inevitably the developer who created code for "Administration of Cat Pictures" checking a role of "Admin" ends up having that role used by the guy granting root access to the filesystem because "Admin" meant different things to both parties. Permissions solve this, if at the expense that a Permission needs to be created for every little thing that a system can do and those Permissions rolled up into Roles that can be assigned to Users.

Notice that I'm not talking at all about specifics of any particular AA scheme. Some providers like Kerberos do authentication but no authorization. PAM is an example of something that can do both, but it locks deployment into Linux hosts and requires extra staff specialization to maintain. There's also issues related to per-service authentication (like OTP, but work consistently over time) and 2FA that real world apps need to handle gracefully (witness the LastPass hack today).

tl;dr: I don't think it's a good idea for Akka to have AAA API built in, rather a separate project that can connect to solid SPIs in Akka and encourage competition so the best solution for specific site requirements can be chosen by those teams.

**Comment by sirthias** [comment]
I agree fully with your assessment.
However, I'm not quite sure about what your suggestion with regard to Akka is.
Are you voting for *removal* of the Basic Auth / OAuth support that's already there?

**Comment by briantopping** [comment]
I think the OAuth support should be removed and the Basic Auth left in. HTTP specs Basic/Digest as a part of the spec so those should be supported, but anything outside of that should be a part of a new project with a decoupled release cycle. The new project might be an Akka subproject such that it can't fall behind WRT to backward incompatible change, but should be separate for the reasons listed above (including hopeful emergence of competing implementations).

**Comment by briantopping** [comment]
To the end of what such an SPI would look like, so far I've been noodling the idea that stream transforms are doing the actual traffic modification (cookies, signatures, etc) and there's a system extension that can interface to these transforms to normalize the AAA information that comes from each of them and present it to the app developer..

**Comment by sirthias** [comment]
Interesting!
Thanks for your perspective, Brian!
I think this is the time where @rkuhn's input is required.

**Comment by ktoso** [comment]
While I agree with all statements made about not locking in or falling behind etc, I'm not quite sure what actually is missing from current infrastructure... We don't implement OAuth any more than just parsing the header, i.e. we do not implement the OAuth protocol, and a library could do that if they want to.

**Comment by rkuhn** [comment]
We’ll see how things evolve: we don’t have the resources to do everything at once, but if there is sufficient need then solutions will be found, either inside or outside of the akka/akka repository.

Nitpick: Accounting is missing from the list :-)

**Comment by ktoso** [comment]
In other words, close since the ticket doesn't seem actionable, at least right now, to me?

**Comment by briantopping** [comment]
Audit ~= Accounting. I'm fine with changing the wording. 

I see both sides of closing the ticket but it seems like there are always tickets that don't have resources. Is it better to have less tickets in the system or more consistency toward an eventual goal? Seems to depend on whether this is a desirable aspect to have in the toolbox or not. Closing the ticket says to me "it's unlikely to ever exist here, please develop elsewhere", which is fine. 

I have the time to develop this if a sponsor could be found.

**Comment by rkuhn** [comment]
I’m not yet sure that it won’t be in scope, so we can leave it open—if an awesome external library solves this then we can close it with a pointer in its direction.

**Comment by ktoso** [comment]
It does not quite fit into Akka HTTP Core, as admitted by a few comments here.
We'll close this for now but would be happy to help review such implementation
